rfcs/20190729-keras-preprocessing.md

```diff
@@ -0,0 +1,732 @@
+        shuffle: Whether to shuffle the data (default: True)
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        follow_links: Whether to follow links inside
+            subdirectories (default: False).
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+```
+
+Notes:
+- Resizing (`target_size`) is moved to constructor (`resize`).
+- `validation_split` is moved from constructor to here.
+
+#### Question: how to support image segmentation in a simple way?
+
+**Requirements:**
+- Image loading and image augmentation should be synced across inputs and targets
+- It should be possible to use different standardization preprocessing (outside of augmentation) across inputs and targets
+
+**Proposal:** (TBD)
+
+```python
+augmenter = ImageDataGenerator(...)
+input_preprocessor = ImageDataGenerator(...)  # or write your own function
+target_preprocessor = ImageDataGenerator(...)  # or write your own function
+
+input_ds = augmenter.dataset_from_directory(input_dir, targets=None, seed=123)
+target_ds = augmenter.dataset_from_directory(target_dir, targets=None, seed=123)  # target_ds should have same structure.
+input_ds = input_ds.map(input_preprocessor.standardize)  # `standardize` could work on tensors.
+target_ds = input_ds.map(target_preprocessor.standardize)
+
+ds = tf.data.Dataset.zip((input_ds, target_ds))
+
+model.fit(ds)
+```
+
+
+
+#### Dataframe methods
+
+```python
+# CURRENT
+
+def flow_from_dataframe(
+    self,
+    dataframe,
+    directory=None,
+    x_col='filename',
+    y_col='class',
+    weight_col=None,
+    target_size=(256, 256),
+    color_mode='rgb',
+    classes=None,
+    class_mode='categorical',
+    batch_size=32,
+    shuffle=True,
+    seed=None,
+    save_to_dir=None,
+    save_prefix='',
+    save_format='png',
+    subset=None,
+    interpolation='nearest',
+    validate_filenames=True,
+    **kwargs):
+```
+
+```python
+# PROPOSED
+
+def generator_from_dataframe(
+    self,
+    dataframe,
+    directory=None,
+    data_column='filename',
+    target_column='class',
+    weight_column=None,
+    color_mode='rgb',
+    class_mode='categorical',
+    batch_size=32,
+    shuffle=True,
+    seed=None,
+    subset=None):
+
+def dataset_from_dataframe(
+    self,
+    dataframe,
+    directory=None,
+    data_column='filename',
+    target_column='class',
+    weight_column=None,
+    color_mode='rgb',
+    class_mode='categorical',
+    batch_size=32,
+    shuffle=True,
+    seed=None,
+    subset=None):
+```
+
+### Utility functions
+
+These are functions located at `affine_transformations.py`.
+
+Proposed changes: API simplification & standardization, and extension of supported types.
+
+- All functions should work on either single images or batches of images
```

Comment:
Do we use the [deprecation procedure](https://github.com/keras-team/keras/issues/12174) for this?

Reply:
As we rename argument, we should keep the old names working (with a deprecation warning).

There is one case where we are changing the semantics of an argument in a non-backwards compatible way: the `preprocessing_function` would now be applied before the transformations rather than after. We could come up with new names in order to avoid breakages. Thoughts?